### PR TITLE
Use System.Text.Json instead of SimpleJson

### PR DESCRIPTION
**Pull request by louis-z**
This entails the following:
- Add specific type converters for `enum`s, `DynamicEnum<T>`, `Sha1`, etc.
- Adapt API so that no JSON serialization types are exposed (as was the case for `GitLabException.ErrorObject`)
 
## Rationale for changing serializers
`System.Text.Json` is a modern serializer that offers much better performances than `SimpleJson`. For instance, using a simple benchmark that deserializes an array of 20 `Project`s, we get the following performances:

image

`System.Text.Json` is 5 times faster and allocates 100 times less memory.

### CODE REVIEW

File: NGitLab.Tests/BranchClientTests.cs [hunk 1]
```diff
@@ -26,7 +26,7 @@ public async Task Test_CommitInfoIsCorrectlyDeserialized()
             Assert.LessOrEqual(7, commit.ShortId.Length);
 
             var fiveMinutesAgo = DateTime.UtcNow - TimeSpan.FromMinutes(5);
-            Assert.Less(fiveMinutesAgo, commit.CreatedAt);
+            Assert.Less(fiveMinutesAgo, commit.CreatedAt.ToUniversalTime());
```
Comment (meziantou) [inline NGitLab.Tests/BranchClientTests.cs]: The current version doesn't return UTC date (`DateTime.Kind`)? Maybe we should fix that in another MR if needed

File: NGitLab/GitLabException.cs [hunk 2]
```diff
@@ -37,8 +37,6 @@ public GitLabException(string message, Exception inner)
         /// </summary>
         public HttpStatusCode StatusCode { get; set; }
 
-        public JsonObject ErrorObject { get; set; }
```
Comment (meziantou) [inline NGitLab/GitLabException.cs:40]: I think you can preserve this property (but changing the return type) by using `JsonExtensionDataAttribute`

File: NGitLab/Impl/Json/EnumConverter.cs [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+using NGitLab.Models;
+
+namespace NGitLab.Impl.Json
+{
+    internal sealed class EnumConverter : JsonConverterFactory
+    {
+        public override bool CanConvert(Type type)
+        {
+            // AccessLevel will be serialized as int, using System.Text.Json's default enum serializer
+            return type.IsEnum && type != typeof(AccessLevel);
```
Comment (meziantou) [inline NGitLab/Impl/Json/EnumConverter.cs]: Can't you use `JsonStringEnumConverter`?

File: NGitLab.Tests/BranchClientTests.cs (hunk 1 above)
Comment (louis-z) [reply NGitLab.Tests/BranchClientTests.cs]: IIRC the result of deserializing `"created_at"` differs
- `SimpleJson` converts it to UTC, while
- `System.Text.Json` keeps it in Local time

File: NGitLab/GitLabException.cs (hunk 2 above)
Comment (louis-z) [reply NGitLab/GitLabException.cs:40]: I always felt it was a mistake to leak out to the public API types that are associated to the serializer. Anything related to the serializer should remain internal.

Furthermore, has `ErrorObject` ever been useful to anyone? I think not, but if I'm wrong, I propose that we make it something more generic. Maybe `IDictionary<string, object>`?

File: NGitLab/Impl/Json/EnumConverter.cs (hunk 3 above)
Comment (louis-z) [reply NGitLab/Impl/Json/EnumConverter.cs]: AFAIK `JsonStringEnumConverter` does not handle the `[EnumMember]` attribute.

File: NGitLab.Tests/BranchClientTests.cs (hunk 1 above)
Comment (meziantou) [reply NGitLab.Tests/BranchClientTests.cs]: GitLab is inconsistent (once again). For an unknown reason it uses "+00:00" instead of "Z" for `created_at`...
So, S.T.Json converts it to a local date.

I would prefer a custom converter if needed to avoid potential issues with date comparison.

The test should be unchanged.

File: NGitLab/GitLabException.cs (hunk 2 above)
Comment (meziantou) [reply NGitLab/GitLabException.cs:40]: We do use this property to extract error details in some cases.
`IDictionary<string, object>` is fine

File: NGitLab/Impl/API.cs [hunk 4]
```diff
@@ -64,7 +64,6 @@ private string OpenPrivateSession()
                 {
                     OriginalCall = new Uri(ex.OriginalCall.OriginalString.Replace(_credentials.Password, hiddenPassword)),
                     StatusCode = ex.StatusCode,
-                    ErrorObject = ex.ErrorObject,
```
Comment (meziantou) [inline NGitLab/Impl/API.cs]: Keep the property and use `IDictionary<string, object?>`

File: NGitLab/Impl/Json/ReflectionExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+                    .Cast<EnumMemberAttribute>()
+                    .FirstOrDefault()?
+                    .Value;
+
+                yield return new EnumMapping((Enum)enumValue, stringValue);
+            }
+        }
+    }
+
+    internal struct EnumMapping
```
Comment (meziantou) [inline NGitLab/Impl/Json/ReflectionExtensions.cs]:
```suggestion
    internal readonly struct EnumMapping
```

File: NGitLab/Impl/Json/Serializer.cs [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+        }
+
+        public static bool TryDeserializeObject(string json, out object obj)
+        {
+            try
+            {
+                obj = Deserialize<object>(json);
+                return true;
+            }
+            catch (Exception /*ex*/)
```
Comment (meziantou) [inline NGitLab/Impl/Json/Serializer.cs]:
```suggestion
            catch
```

File: NGitLab/Models/MergeRequestApprove.cs [hunk 7]
```diff
@@ -9,12 +10,14 @@ public class MergeRequestApprove
...
         /// </summary>
         [DataMember(Name = "sha")]
+        [JsonPropertyName("sha")]
         public string Sha { get; set; }
 
         /// <summary>
         /// (optional) - Current user’s password. Required if Require user password to approve is enabled in the project settings.
         /// </summary>
         [DataMember(Name = "approval_password ")]
+        [JsonPropertyName("approval_password ")]
```
Comment (meziantou) [inline NGitLab/Models/MergeRequestApprove.cs]:
```suggestion
        [JsonPropertyName("approval_password")]
```

File: NGitLab/PublicAPI.Unshipped.txt [hunk 8]
```diff
@@ -2818,6 +2793,7 @@ NGitLab.Models.UserToken.CreatedAt.get -> System.DateTime
 NGitLab.Models.UserToken.CreatedAt.set -> void
 NGitLab.Models.UserToken.ExpiresAt.get -> System.DateTime?
 NGitLab.Models.UserToken.ExpiresAt.set -> void
+NGitLab.Models.UserToken.ExpiresAtStr.get -> string
```
Comment (meziantou) [inline NGitLab/PublicAPI.Unshipped.txt]: I think we can avoid these new props by using a custom converter for DateTime

File: NGitLab/PublicAPI.Unshipped.txt [hunk 9]
```diff
@@ -2773,22 +2748,22 @@ NGitLab.Models.User.CanCreateGroup -> bool
 NGitLab.Models.User.CanCreateProject -> bool
 NGitLab.Models.User.ColorSchemeId -> int
 NGitLab.Models.User.CommitEmail -> string
-NGitLab.Models.User.ConfirmedAt -> System.DateTime
+NGitLab.Models.User.ConfirmedAt -> System.DateTime?
 NGitLab.Models.User.CreatedAt -> System.DateTime
-NGitLab.Models.User.CurrentSignIn -> System.DateTime
+NGitLab.Models.User.CurrentSignIn -> System.DateTime?
```
Comment (meziantou) [inline NGitLab/PublicAPI.Unshipped.txt]: Could we implement a custom converter to set the prop to `DateTime.MinValue` (current behavior) when the json prop is null?

File: NGitLab/PublicAPI.Unshipped.txt [hunk 10]
```diff
@@ -2550,14 +2525,14 @@ NGitLab.Models.ResourceLabelEventAction.Add = 0 -> NGitLab.Models.ResourceLabelE
...
-NGitLab.Models.Runner.ContactedAt -> System.DateTime
+NGitLab.Models.Runner.ContactedAt -> System.DateTime?
 NGitLab.Models.Runner.Description -> string
 NGitLab.Models.Runner.Id -> int
 NGitLab.Models.Runner.IpAddress -> string
 NGitLab.Models.Runner.IsShared -> bool
 NGitLab.Models.Runner.Locked -> bool
 NGitLab.Models.Runner.Name -> string
-NGitLab.Models.Runner.Online -> bool
+NGitLab.Models.Runner.Online -> bool?
```
Comment (meziantou) [inline NGitLab/PublicAPI.Unshipped.txt]: Could we use a custom converter to avoid the breaking change?

File: NGitLab/PublicAPI.Unshipped.txt [hunk 11]
```diff
@@ -1358,7 +1333,7 @@ NGitLab.Models.Event.TargetIId.get -> long?
 NGitLab.Models.Event.TargetIId.set -> void
 NGitLab.Models.Event.TargetTitle.get -> string
 NGitLab.Models.Event.TargetTitle.set -> void
-NGitLab.Models.Event.TargetType.get -> NGitLab.DynamicEnum<NGitLab.Models.EventTargetType>
+NGitLab.Models.Event.TargetType.get -> NGitLab.DynamicEnum<NGitLab.Models.EventTargetType>?
```
Comment (meziantou) [inline NGitLab/PublicAPI.Unshipped.txt]: I think a custom converter could map `null` to `EventTargetType.None` to avoid the breaking change

File: NGitLab.Tests/BranchClientTests.cs (hunk 1 above)
Comment (louis-z) [reply NGitLab.Tests/BranchClientTests.cs]: Added a custom converter & put back test as it was before.

File: NGitLab/GitLabException.cs (hunk 2 above)
Comment (louis-z) [reply NGitLab/GitLabException.cs:40]: Done

File: NGitLab/PublicAPI.Unshipped.txt (hunk 8 above)
Comment (louis-z) [reply NGitLab/PublicAPI.Unshipped.txt]: Done

File: NGitLab/PublicAPI.Unshipped.txt (hunk 11 above)
Comment (louis-z) [reply NGitLab/PublicAPI.Unshipped.txt]: Done

File: NGitLab/Impl/Json/BooleanConverter.cs [hunk 12]
```diff
@@ -0,0 +1,22 @@
+﻿using System;
+using System.Text.Json;
+using System.Text.Json.Serialization;
+
+namespace NGitLab.Impl.Json
+{
+    internal sealed class BooleanConverter : JsonConverter<bool>
```
Comment (meziantou) [inline NGitLab/Impl/Json/BooleanConverter.cs:7]: Could you add some tests for these converters?

File: NGitLab/Impl/Json/Int64Converter.cs [hunk 13]
```diff
@@ -0,0 +1,33 @@
+﻿using System;
+using System.Text.Json;
+using System.Text.Json.Serialization;
+
+namespace NGitLab.Impl.Json
+{
+    internal sealed class Int64Converter : JsonConverter<long>
```
Comment (meziantou) [inline NGitLab/Impl/Json/Int64Converter.cs:8]: Could you add a test to validate this converter is not used for `long?`?

File: NGitLab/Impl/Json/Int64Converter.cs [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+                return 0;
+
+            if (reader.TokenType == JsonTokenType.Number)
+                return reader.GetInt64();
+
+            // Change.AMode & BMode are deserialized from strings...
+            if (reader.TokenType == JsonTokenType.String)
+            {
+                var stringValue = reader.GetString();
+                if (long.TryParse(stringValue, out var value))
```
Comment (meziantou) [inline NGitLab/Impl/Json/Int64Converter.cs]: CultureInvariant

File: NGitLab/Impl/Json/BooleanConverter.cs (hunk 12 above)
Comment (louis-z) [reply NGitLab/Impl/Json/BooleanConverter.cs:7]: The converters were added to make existing tests pass. So, the converter code is already validated (and without it tests would fail). Do you still think we need to add more tests?

File: NGitLab/Impl/Json/Int64Converter.cs (hunk 13 above)
Comment (louis-z) [reply NGitLab/Impl/Json/Int64Converter.cs:8]: According to the [System.Text.Json documentation](https://docs.microsoft.com/en-us/dotnet/standard/serialization/system-text-json-converters-how-to?pivots=dotnet-6-0#handle-null-values), this custom converter will **not** be used for `Nullable<long>`:

```
By default, the serializer handles null values as follows:
- For reference types and Nullable<T> types:
  - It does not pass null to custom converters on serialization.
  - It does not pass JsonTokenType.Null to custom converters on deserialization.
  - It returns a null instance on deserialization.
  - It writes null directly with the writer on serialization.
- For non-nullable value types:
  - It passes JsonTokenType.Null to custom converters on deserialization. (If no custom converter is available, a JsonException exception is thrown by the internal converter for the type.)
```

So unless you think the documentation is lying 😁, I don't see the need for such a test.

What do you think?

File: NGitLab/Impl/Json/BooleanConverter.cs (hunk 12 above)
Comment (meziantou) [reply NGitLab/Impl/Json/BooleanConverter.cs:7]: You don't need to validate them all, but I would prefer some tests to validate it correctly handle `null` values and also the parsing of datetime in different format (and that they are all UTC).

File: NGitLab/Impl/Json/Int64Converter.cs (hunk 13 above)
Comment (meziantou) [reply NGitLab/Impl/Json/Int64Converter.cs:8]: Let's trust the documentation!

File: NGitLab/Impl/Json/Int64Converter.cs (hunk 14 above)
Comment (louis-z) [reply NGitLab/Impl/Json/Int64Converter.cs]: Done

File: NGitLab/Impl/Json/BooleanConverter.cs (hunk 12 above)
Comment (louis-z) [reply NGitLab/Impl/Json/BooleanConverter.cs:7]: I added some tests. Let me know if you feel there should be more.

File: NGitLab.Tests/Impl/JsonConverterTests.cs [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+
+        [TestCase("2022-01-12")]
+        [TestCase("2022-01-12T22:49:21.552Z")]
+        [TestCase("2022-01-12T22:49:21.552+00:00")]
+        public void Test_DeserializeStringToDateTime_SupportsMultipleFormats(string input)
+        {
+            var json = $@"{{ ""a_date_time"": ""{input}"" }}";
+            var obj = Serializer.Deserialize<MyDataContract>(json);
+            Assert.NotNull(obj);
+            Assert.AreEqual(DateTimeKind.Utc, obj.SomeDateTime.Kind);
```
Comment (meziantou) [inline NGitLab.Tests/Impl/JsonConverterTests.cs]: Should it be unknown for `[TestCase("2022-01-12")]`?

Comment (louis-z) [reply NGitLab.Tests/Impl/JsonConverterTests.cs]: Yeah, I guess that would make sense. I'll revisit the code...

Comment (louis-z) [reply NGitLab.Tests/Impl/JsonConverterTests.cs]: Done.

Also, I removed unnecessary `DateTime` format strings, i.e. those without which the tests still pass.
